devito/yask/transformer.py

```diff
@@ -130,49 +135,54 @@ def nary2binary(args, op):
         if function.is_Constant:
             if function not in mapper:
                 mapper[function] = yc_soln.new_grid(function.name, [])
-            return mapper[function].new_relative_grid_point([])
-        elif not function.is_Dimension:
+            return mapper[function].new_grid_point([])
+        elif function.is_Dimension:
+            if expr.is_Time:
+                return nfac.new_step_index(expr.name)
+            elif expr.is_Space:
+                return nfac.new_domain_index(expr.name)
+            else:
+                return nfac.new_misc_index(expr.name)
+        else:
             # A DSE-generated temporary, which must have already been
             # encountered as a LHS of a previous expression
             assert function in mapper
             return mapper[function]
     elif expr.is_Indexed:
+        # Create a YASK compiler grid if it's the first time we encounter a Function
         function = expr.base.function
         if function not in mapper:
-            if function.is_TimeFunction:
-                dimensions = [nfac.new_step_index(function.indices[0].name)]
-                dimensions += [nfac.new_domain_index(i.name)
-                               for i in function.indices[1:]]
-            else:
-                dimensions = [nfac.new_domain_index(i.name)
-                              for i in function.indices]
+            dimensions = [make_yask_ast(i, yc_soln, mapper) for i in function.indices]
             mapper[function] = yc_soln.new_grid(function.name, dimensions)
-        # Detect offset from dimension. E.g., in `[x+3,y+4]`, detect `[3,4]`
+        # Convert the Indexed into a YASK grid access
         indices = []
-        for i, j in zip(expr.indices, function.indices):
+        for i in expr.indices:
             if isinstance(i, LoweredDimension):
-                access = i.origin
+                indices.append(make_yask_ast(i.origin, yc_soln, mapper))
+            elif i.is_integer:
+                # A YASK misc dimension causes this
+                indices.append(make_yask_ast(i, yc_soln, mapper))
             else:
-                # SubDimension require this
+                # We must always use the parent ("main") dimension when creating
+                # YASK expressions
                 af = split_affine(i)
                 dim = af.var.parent if af.var.is_Derived else af.var
-                access = dim + af.shift
-            indices.append(int(access - j))
-        return mapper[function].new_relative_grid_point(indices)
```

Comment:
I just thought of another case that might need handling here, if it doesn't do it already (I have to run shortly but can test later).

The BCs often use the velocity value just inside the boundary, so they will have an expression like.

`u[t+1, xleft, y, z] = some_function_of(vel[x_m+3, y, z], u[t, ....], u[t-1, ....])`

These expressions involving `x_m` and `x_M` may need handling here just like in the loop case in the previous PR.  (in fact, if they are handled here you might even just be able to call `make_yask_ast` from that code.

Reply:
right, but this opens another question. How do we propagate the runtime value ``x_m`` to YASK? (@chuckyount ) perhaps it's worth discussing this on Slack

Reply:
dumb me, just use the usual ``FIRST_INDEX(x)`` or whatever is called, I guess?

Reply:
The PR is already quite big, and there are others coming along. I suggest to leave this for another day (until we have a test)